### PR TITLE
refactor(providers): unify OpenAI schema converter and preserve dropped JSON-schema keywords (Fixes #2508)

### LINKED ISSUE #2508: Unify duplicated OpenAI schema converter into shared util and fix dropped JSON-schema keywords (anyOf/format/pattern/etc)
## Summary

The OpenAI tool-schema converter is duplicated across the classic and vercel providers, and both copies share the same latent correctness bug: property-level JSON-schema keywords are silently dropped. We should (1) extract a single shared converter utility, and (2) fix the extracted util so it preserves the currently-dropped keywords.

## Duplication (DRY)

These two files are 206 lines each and byte-identical except for cosmetic differences (doc comments, logger namespace, exported function name convertToolsToOpenAI vs convertToolsToOpenAIVercel, and one trivial detail: classic uses description: decl.description ?? '' while vercel uses description: decl.description):

- packages/providers/src/openai/schemaConverter.ts
- packages/providers/src/openai-vercel/schemaConverter.ts

The core logic (convertSchemaToOpenAI, convertPropertySchema, convertProperties, normalizeType, toNumber, isSchemaObject) is identical in both. This is copy-paste that will drift.

## Latent correctness bug (in BOTH copies)

convertPropertySchema() is a whitelist converter. For each property it only copies:

- type, description, enum, items, nested properties (+ nested required), minimum, maximum, minLength, maxLength, default

It silently DROPS these property-level keywords:

- anyOf, oneOf, allOf
- $ref
- const
- format
- pattern
- additionalProperties (only handled at the top-level object in convertSchemaToOpenAI, not on nested properties)

Any tool whose parameter uses a union type (anyOf/oneOf), a $ref, or format/pattern/const constraints will have that structure quietly stripped before being sent to the model. Our current built-in tools (exa_web_search, glob, read_file) happen to use none of these, so nothing is lost today — but this is a correctness landmine for any future tool or MCP-provided tool schema.

Note: this was discovered while investigating why a local Ollama model was calling tools with wrong arguments. In that specific case the schemas were clean (the model was just weak), but the whitelist gap is real and provider-wide.

## Proposed solution

1. Extract the converter into a single shared module (e.g. packages/providers/src/openai-shared/schemaConverter.ts) exporting convertSchemaToOpenAI and a convertToolDeclarations helper.
2. Have both the classic openai provider and the openai-vercel provider import the shared util. Keep only the thin per-provider wrapper if needed (the convertToolsToOpenAI vs convertToolsToOpenAIVercel difference is just a name; the description ?? '' vs description difference can be a parameter or normalized to one behavior).
3. Fix the extracted converter to PRESERVE unknown/unhandled JSON-schema keywords instead of dropping them. Keep the existing normalizations (uppercase to lowercase type, ensure required is always an array, numeric-string coercion) but pass through anyOf, oneOf, allOf, $ref, const, format, pattern, and property-level additionalProperties. Recurse into anyOf/oneOf/allOf sub-schemas so their types are normalized too.
4. Add tests covering: a property with anyOf, a property with format/pattern, nested additionalProperties, and a $ref, asserting they survive conversion; plus the existing normalizations still apply.

## Acceptance criteria

- Exactly one implementation of the OpenAI schema converter; both providers use it.
- A tool parameter using anyOf/oneOf/format/pattern/const/$ref/nested additionalProperties is preserved end-to-end (verified in the request sent to the model).
- Existing behavior preserved: type normalized to lowercase, required always present as an array, numeric string constraints coerced to numbers.
- No behavioral regression for the current built-in tools.

## Guardrails

Do not loosen lint/complexity rules and do not add eslint-disable / ts-ignore / ts-expect-error suppressions. This is a DRY/modularization win; keep the shared util within existing complexity limits and fix the underlying issue rather than silencing rules.

### PR DESCRIPTION
**Pull request by acoliver**
## Summary

Fixes #2508

The OpenAI tool-schema converter was duplicated byte-for-byte across the `openai` and `openai-vercel` providers (~206 lines each), and both copies shared a latent correctness bug: `convertPropertySchema` was a whitelist converter that **silently dropped** property-level JSON-schema keywords — `anyOf`, `oneOf`, `allOf`, `$ref`, `const`, `format`, `pattern`, and nested `additionalProperties`. Any tool whose parameters used a union type, a reference, or validation constraints had that structure quietly stripped before reaching the model.

This PR (1) extracts a single shared converter and (2) fixes the bug so dropped keywords are preserved end-to-end.

## Changes

- **New shared module** `packages/providers/src/openai-shared/schemaConverter.ts` — the single source of truth. Exports `convertSchemaToOpenAI`, `convertToolDeclarations`, and the shared types (`OpenAIFunctionParameters`, `OpenAIPropertySchema`).
- **`openai/schemaConverter.ts`** — reduced to a thin wrapper (`convertToolsToOpenAI`) that delegates to the shared converter with the classic description strategy (always a string) and preserves the classic logger namespace / log message.
- **`openai-vercel/schemaConverter.ts`** — reduced to a thin wrapper (`convertToolsToOpenAIVercel`) with the preserve description strategy and the vercel logger namespace / log message.
- **New tests** `packages/providers/src/openai-shared/__tests__/schemaConverter.test.ts` (25 behavioral tests).

## The bug fix

The whitelist `convertPropertySchema` is replaced by a recursive passthrough normalizer:

- Handled keywords (`type`, `description`, `enum`, `items`, `properties`, `required`, `minimum`/`maximum`/`minLength`/`maxLength`, `default`) are normalized exactly as before.
- **Every other keyword is preserved** and, where the value is a schema (or array of schemas), normalized recursively so nested types are lowercased too. This covers `anyOf`/`oneOf`/`allOf` (recursed element-wise), `$ref`, `const`, `format`, `pattern`, and object-valued `additionalProperties`.
- Plain data objects embedded in keywords like `const`/`default`/`examples` are detected (they lack schema-signaling keys) and passed through **unchanged**, so they are not corrupted by schema normalization (e.g. a `const: { userId: 123 }` is not given a spurious `type`).
- Top-level keywords beyond `properties`/`required`/`additionalProperties` are now also passed through (e.g. root `description`, `$schema`).

## Acceptance criteria

- [x] Exactly one implementation of the OpenAI schema converter; both providers use it.
- [x] A tool parameter using `anyOf`/`oneOf`/`allOf`/`format`/`pattern`/`const`/`$ref`/nested-`additionalProperties` is preserved end-to-end (covered by tests and verified by direct invocation).
- [x] Existing behavior preserved: type normalized to lowercase, `required` always present as an array, numeric-string constraints coerced to numbers, the Gemini numeric type enum mapped, the missing-`parametersJsonSchema` error message unchanged, empty tools return `undefined`.
- [x] No behavioral regression for the current built-in tools (full providers suite passes).
- [x] `openai-responses` converter intentionally left untouched (different output shape; out of scope).

## Per-provider differences preserved

The wrappers keep the original cosmetic differences so consumer code is unaffected:
- Function names: `convertToolsToOpenAI` vs `convertToolsToOpenAIVercel`.
- Tool type shapes: classic `function.parameters` required with `description: string`; vercel both optional.
- Description handling: classic coerces missing to empty string; vercel preserves `undefined`.
- Logger namespaces and log messages.

## Verification

- `npm run typecheck` — clean across all packages.
- `npm run lint` (providers package) + `npm run lint:eslint-guard` — clean, no rules loosened, no suppressions.
- Providers test suite — 1079 passed, 3 skipped (including 25 new shared-converter tests).
- `npm run format` / `npm run build` — clean.
- Open Code Review (ocr) run with `--timeout 20`; all High/Medium findings remediated (top-level keyword passthrough, non-schema object corruption in const, defensive empty-results guard, expanded test coverage).

## Guardrails

No lint/complexity rules were loosened and no eslint-disable / ts-ignore / ts-expect-error suppressions were added. The shared module stays within the project's complexity limits via small extracted helpers.

### CODE REVIEW

File: packages/providers/src/openai-shared/__tests__/schemaConverter.test.ts [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+      },
+      required: [],
+    };
+
+    const result = convertSchemaToOpenAI(schema);
+    const config = result.properties.config as Record<string, unknown>;
+
+    expect(config.const).toStrictEqual({ userId: 123, active: true });
+    expect(config.type).toBe('string');
+  });
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/openai-shared/__tests__/schemaConverter.test.ts]: Valid catch — the test name 'no type injection' contradicted the assertion. Fixed in 124cf6e96: renamed the test to describe the actual guarantee (the const object value is preserved verbatim, uncorrupted by schema normalization) and dropped the misleading assertion. Resolving.

File: packages/providers/src/openai-shared/schemaConverter.ts [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+      }
+    }
+  } else if (isSchemaObject(node.items)) {
+    result.items = normalizeSchemaNode(node.items);
+  }
+
+  if (node.properties != null && typeof node.properties === 'object') {
+    result.properties = convertProperties(
+      node.properties as Record<string, unknown>,
+    );
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/openai-shared/schemaConverter.ts:245]: Two corrections: (1) '!= null is prohibited' is factually incorrect — this project's ESLint config has no such rule, and 'npm run lint' passes clean on this file. (2) This exact line ('node.properties != null && typeof node.properties === object') is pre-existing behavior copied verbatim from both original converters (openai/schemaConverter.ts and openai-vercel/schemaConverter.ts). The issue explicitly requires preserving existing normalizations and avoiding behavioral regression, so this is intentionally unchanged and out of scope for this DRY/keyword-passthrough PR. Resolving as out-of-scope/pre-existing.

File: packages/providers/src/openai-shared/schemaConverter.ts [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+  const result: OpenAIFunctionParameters = {
+    type: 'object',
+    properties: {},
+    required: [],
+  };
+
+  if (input.properties != null && typeof input.properties === 'object') {
+    result.properties = convertProperties(
+      input.properties as Record<string, unknown>,
+    );
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/openai-shared/schemaConverter.ts:383]: Same as the other comment on line 219: this '!= null' check is pre-existing behavior preserved verbatim from the original converters, and the project's ESLint config does not prohibit it (lint passes clean). Out of scope for this PR. Resolving.

File: packages/providers/src/openai-shared/schemaConverter.ts [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+  node: Record<string, unknown>,
+  result: OpenAIPropertySchema,
+): void {
+  if (typeof node.description === 'string') {
+    result.description = node.description;
+  }
+
+  if (Array.isArray(node.enum)) {
+    result.enum = node.enum.map((v) => String(v));
+  }
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/openai-shared/schemaConverter.ts:229]: Valid observation in general, but this String(v) enum coercion is pre-existing behavior copied verbatim from both original converters. The issue scope is specifically (1) DRY consolidation and (2) preserving currently-dropped keywords, and it explicitly requires keeping the existing normalizations unchanged ('Keep the existing normalizations'). Changing enum-value coercion would be a behavioral change beyond this PR's scope. Resolving as out-of-scope/pre-existing.

File: packages/providers/src/openai-shared/schemaConverter.ts [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+ * corrupted by schema normalization.
+ */
+function isLikelySchemaNode(node: Record<string, unknown>): boolean {
+  for (const key of Object.keys(node)) {
+    if (SCHEMA_VALUE_KEYS.has(key)) {
+      return true;
+    }
+  }
+  return false;
+}
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/openai-shared/schemaConverter.ts]: Valid and important catch — reproduced the corruption before fixing (a const object using schema-ish key names like { type, properties } had its properties array replaced with an empty object). Fixed in 64cf36422 by replacing the isLikelySchemaNode key-name heuristic with positional keyword normalization: anyOf/oneOf/allOf, not/if/then/else, and additionalProperties are normalized by their known JSON-schema positions, while const/default/examples/format/pattern/$ref/title/... are copied verbatim and never recursed into. Added regression tests (const and default objects with schema-ish keys). Resolving.

File: packages/providers/src/openai-shared/schemaConverter.ts [hunk 6]
```diff
@@ -0,0 +1,438 @@
...
+  items?: OpenAIPropertySchema;
+  properties?: Record<string, OpenAIPropertySchema>;
+  required?: string[];
+  minimum?: number;
+  maximum?: number;
+  minLength?: number;
+  maxLength?: number;
+  default?: unknown;
+  [key: string]: unknown;
+}
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/openai-shared/schemaConverter.ts:58]: Valid JSON-schema observation, but this is pre-existing behavior intentionally left unchanged and out of scope for this PR. Verified against the original converters: both openai/schemaConverter.ts and openai-vercel/schemaConverter.ts used isSchemaObject() for items and typeof === 'object' for property values, so boolean schemas (items: true, or a boolean property value) were already dropped BEFORE this PR. The issue (#2508) scope is specifically (1) DRY consolidation and (2) preserving the named dropped keywords (anyOf/oneOf/allOf/$ref/const/format/pattern/nested additionalProperties), with the explicit guardrail 'no behavioral regression for the current built-in tools' (none of which use boolean schemas). Properly supporting boolean schemas requires widening OpenAIPropertySchema.type from string to string | boolean, which ripples into the provider wrappers and their consumers (e.g. vercelModelClient.ts casts parameters to JSONSchema7) — a type-system change beyond this PR. This refactor does not make boolean handling any worse than before. Recommend tracking as a separate issue. Resolving as out-of-scope/pre-existing.

File: packages/providers/src/openai-shared/schemaConverter.ts [hunk 7]
```diff
@@ -0,0 +1,438 @@
...
+  if (Array.isArray(node.items)) {
+    if (node.items.length > 0) {
+      const firstItem = node.items[0];
+      if (isSchemaObject(firstItem)) {
+        result.items = normalizeSchemaNode(firstItem);
+      }
+    }
+  } else if (isSchemaObject(node.items)) {
+    result.items = normalizeSchemaNode(node.items);
+  }
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/openai-shared/schemaConverter.ts:240]: Same boolean-schema concern as the comment on the OpenAIPropertySchema interface (line 58). Pre-existing behavior (the original converters also dropped boolean items/properties via isSchemaObject), and out of scope for this PR per the issue guardrails. See my reply there for the full reasoning. Resolving as out-of-scope/pre-existing.

File: packages/providers/src/openai-shared/schemaConverter.ts [hunk 8]
```diff
@@ -0,0 +1,438 @@
...
+  const result: Record<string, OpenAIPropertySchema> = {};
+
+  for (const [key, value] of Object.entries(properties)) {
+    if (isSchemaObject(value)) {
+      result[key] = normalizeSchemaNode(value);
+    }
+  }
+
+  return result;
+}
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/openai-shared/schemaConverter.ts:321]: Same boolean-schema concern — see my reply on the OpenAIPropertySchema interface comment. convertProperties only processes object property values, matching the original converters behavior; boolean property values were already dropped before this PR. Out of scope; resolving as out-of-scope/pre-existing.

File: packages/providers/src/openai-shared/schemaConverter.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/openai-shared/schemaConverter.ts:240]: Duplicate of the items boolean-schema comment above. Pre-existing (original converters used isSchemaObject for items, dropping boolean items: true), out of scope for this PR. Resolving as out-of-scope/pre-existing.

File: packages/providers/src/openai-shared/schemaConverter.ts [hunk 9]
```diff
@@ -0,0 +1,457 @@
...
+  if (node.properties != null && typeof node.properties === 'object') {
+    result.properties = convertProperties(
+      node.properties as Record<string, unknown>,
+    );
+    if (Array.isArray(node.required)) {
+      result.required = node.required.map((r) => String(r));
+    } else if (result.type === 'object') {
+      result.required = [];
+    }
+  }
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/openai-shared/schemaConverter.ts:257]: Valid — verified this was pre-existing behavior (the original convertPropertySchema also gated required inside the properties block), but it aligns with this PR's spirit of not dropping valid schema info, so I fixed it. In 099c55ad9 the nested required handling is moved outside the properties block (with a regression test asserting a nested object schema with required and no inline properties now propagates required). Resolving.

File: packages/providers/src/openai-shared/schemaConverter.ts [hunk 10]
```diff
@@ -0,0 +1,457 @@
...
+    return normalizeSchemaNode(value);
+  }
+  return value;
+}
+
+/**
+ * Normalize a passthrough keyword value by its known JSON-schema position.
+ * - anyOf/oneOf/allOf hold arrays of sub-schemas → each element normalized.
+ * - anyOf/oneOf/allOf hold arrays of sub-schemas → each element normalized.
+ * - not/if/then/else hold a single sub-schema → normalized.
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/openai-shared/schemaConverter.ts:193]: Fixed in 099c55ad9 — removed the duplicated bullet. Resolving.

File: packages/providers/src/openai-shared/__tests__/schemaConverter.test.ts [hunk 11]
```diff
@@ -0,0 +1,468 @@
...
+      description: 'File path',
+    });
+    expect(result.properties.offset).toStrictEqual({
+      type: 'integer',
+      description: 'Line offset',
+    });
+  });
+});
+
+describe('convertToolDeclarations — description strategy', () => {
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/openai-shared/__tests__/schemaConverter.test.ts:467]: Good suggestion — added a positive end-to-end integration test for convertToolDeclarations in 099c55ad9. It passes a complex nested parametersJsonSchema (nested objects, array items, and a union) and asserts the full output structure including type normalization, required propagation, and union preservation. Resolving.

File: packages/providers/src/openai-shared/__tests__/schemaConverter.test.ts [hunk 12]
```diff
@@ -0,0 +1,468 @@
...
+    };
+
+    const result = convertSchemaToOpenAI(schema);
+
+    expect(result.description).toBe('root description');
+    expect(result.$schema).toBe('http://json-schema.org/draft-07/schema#');
+  });
+});
+
+describe('convertSchemaToOpenAI — existing normalizations still apply', () => {
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/openai-shared/__tests__/schemaConverter.test.ts:333]: Added in 099c55ad9 — test 'normalizes uppercase types recursively inside array items' covers { type: ARRAY, items: { type: STRING } } and asserts both the array and item types are lowercased. Resolving.

File: packages/providers/src/openai-shared/__tests__/schemaConverter.test.ts [hunk 13]
```diff
@@ -0,0 +1,468 @@
...
+
+    const result = convertSchemaToOpenAI(schema);
+    const email = result.properties.email as Record<string, unknown>;
+
+    expect(email.type).toBe('string');
+    expect(email.format).toBe('email');
+    expect(email.pattern).toBe('^.+@.+$');
+  });
+
+  it('preserves const', () => {
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/openai-shared/__tests__/schemaConverter.test.ts:174]: Added in 099c55ad9 — test 'preserves primitive const/default values that coincidentally match keyword names' covers const: 'string' and default: 'object' and asserts they survive verbatim. Resolving.

File: packages/providers/src/openai-shared/__tests__/schemaConverter.test.ts [hunk 14]
```diff
@@ -0,0 +1,468 @@
...
+    const result = convertSchemaToOpenAI(schema);
+    const count = result.properties.count;
+
+    expect(count.minimum).toBe(5);
+    expect(count.maximum).toBe(10);
+    expect(count.minLength).toBe(1);
+    expect(count.maxLength).toBe(3);
+  });
+
+  it('maps the Gemini numeric Type enum to a lowercase string', () => {
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/openai-shared/__tests__/schemaConverter.test.ts]: Added in 099c55ad9 — the Gemini Type enum test now covers all values 1-6 (string/number/integer/boolean/array/object) plus a separate test for an out-of-range value (999) asserting the 'string' fallback. Resolving.

File: packages/providers/src/openai-shared/schemaConverter.ts [hunk 15]
```diff
@@ -0,0 +1,460 @@
...
+    if (!toolGroup.functionDeclarations) {
+      continue;
+    }
+
+    for (const decl of toolGroup.functionDeclarations) {
+      if (!isSchemaObject(decl.parametersJsonSchema)) {
+        throw new Error(
+          `Tool "${decl.name}" is missing parametersJsonSchema — legacy schema fallback has been removed. ` +
+            `Ensure all tool declarations provide parametersJsonSchema at construction time.`,
+        );
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/openai-shared/schemaConverter.ts:458]: Valid documentation point. The throw is intentional fail-loud design (the error message itself documents the contract: "Ensure all tool declarations provide parametersJsonSchema at construction time"). I kept the throw as-is rather than adding a silent fallback, since masking a missing schema would send a malformed (empty) tool definition to the model and be much harder to debug. I added a `@throws` JSDoc to `convertToolDeclarations` in 102fe4755 so external callers understand the contract before consuming the exported module.

File: packages/providers/src/openai-shared/__tests__/schemaConverter.test.ts [hunk 16]
```diff
@@ -0,0 +1,604 @@
...
+  it('throws when parametersJsonSchema is missing', () => {
+    const badTools = [
+      {
+        functionDeclarations: [{ name: 'no_schema', description: 'no schema' }],
+      },
+    ];
+
+    expect(() =>
+      convertToolDeclarations(badTools, { descriptionStrategy: 'preserve' }),
+    ).toThrow('Tool "no_schema" is missing parametersJsonSchema');
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/openai-shared/__tests__/schemaConverter.test.ts]: Valid — decoupled the assertion from the exact wording in 102fe4755. It now matches only the stable tool-name token (`no_schema`), which is the meaningful invariant (verifies the offending tool is correctly identified) without coupling to the rest of the message text. The same fix applies to the sibling assertion below.

File: packages/providers/src/openai-shared/__tests__/schemaConverter.test.ts [hunk 17]
```diff
@@ -0,0 +1,604 @@
...
+            description: 'null schema',
+            parametersJsonSchema: null,
+          },
+        ],
+      },
+    ];
+
+    expect(() =>
+      convertToolDeclarations(badTools, { descriptionStrategy: 'preserve' }),
+    ).toThrow('Tool "null_schema" is missing parametersJsonSchema');
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/openai-shared/__tests__/schemaConverter.test.ts]: Fixed in 102fe4755 — this assertion now matches only the stable tool-name token (`null_schema`), consistent with the sibling test above.

File: packages/providers/src/openai-shared/__tests__/schemaConverter.test.ts [hunk 18]
```diff
@@ -0,0 +1,604 @@
...
+  });
+
+  it('returns undefined when there are no tools', () => {
+    expect(
+      convertToolDeclarations(undefined, {
+        descriptionStrategy: 'always-string',
+      }),
+    ).toBeUndefined();
+  });
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/openai-shared/__tests__/schemaConverter.test.ts:537]: Good suggestions — added all three edge-case tests in 102fe4755:
(1) Empty `functionDeclarations` array → asserts the function returns `undefined` (the converted-length check short-circuits).
(2) Declaration missing `name` → documents current passthrough behavior (the resulting tool has `function.name === undefined`); I kept this as a behavior-documenting test rather than adding a new throw, since adding a name-validation throw is a behavioral change outside this refactor scope and the existing built-in tool declarations always supply a name.
(3) Non-array `functionDeclarations` value → asserts it throws (the `for...of` iterates the string and each element triggers the missing-schema throw).

File: packages/providers/src/openai-shared/schemaConverter.ts [hunk 19]
```diff
@@ -0,0 +1,464 @@
...
+  const result: Record<string, OpenAIPropertySchema> = {};
+
+  for (const [key, value] of Object.entries(properties)) {
+    if (isSchemaObject(value)) {
+      result[key] = normalizeSchemaNode(value);
+    }
+  }
+
+  return result;
+}
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/openai-shared/schemaConverter.ts:321]: Valid observation about silent skipping, but this is pre-existing behavior preserved verbatim from both original converters, and it is correct per the JSON Schema spec: each `properties` entry must itself be a schema object, so a non-object value (string/number/boolean/null) is malformed input, not data to preserve. I added a JSDoc on `convertProperties` documenting that non-schema values are intentionally skipped and why.

File: packages/providers/src/openai-shared/schemaConverter.ts [hunk 20]
```diff
@@ -0,0 +1,464 @@
...
+function toNumber(value: unknown): number | undefined {
+  if (typeof value === 'number') {
+    return value;
+  }
+  if (typeof value === 'string') {
+    const num = parseFloat(value);
+    return isNaN(num) ? undefined : num;
+  }
+  return undefined;
+}
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/openai-shared/schemaConverter.ts:131]: Valid in the abstract, but this `parseFloat` coercion is pre-existing behavior copied verbatim from both original converters. Switching to `Number()` would turn partially-numeric constraint strings (e.g. "10px") into `NaN` -> `undefined`, silently dropping the constraint entirely — a behavioral regression. I added a comment documenting why the tolerant coercion is intentional and preserved.

File: packages/providers/src/openai-shared/schemaConverter.ts [hunk 21]
```diff
@@ -0,0 +1,464 @@
...
+      2: 'number',
+      3: 'integer',
+      4: 'boolean',
+      5: 'array',
+      6: 'object',
+    };
+    return typeMap[type] ?? 'string';
+  }
+  return 'string';
+}
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/openai-shared/schemaConverter.ts:113]: Pre-existing behavior: the original converters used the same tolerant `typeMap[type] ?? "string"` fallback. Throwing on an unknown enum value would be a behavioral change (and a runtime crash) outside the scope of this DRY/preservation refactor. I added a comment documenting that the `string` fallback is the intentional tolerant default for malformed input.

File: packages/providers/src/openai-shared/schemaConverter.ts [hunk 22]
```diff
@@ -0,0 +1,464 @@
...
+  node: Record<string, unknown>,
+  result: OpenAIPropertySchema,
+): void {
+  if (typeof node.description === 'string') {
+    result.description = node.description;
+  }
+
+  if (Array.isArray(node.enum)) {
+    result.enum = node.enum.map((v) => String(v));
+  }
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/openai-shared/schemaConverter.ts:229]: Agreed the intent should be documented. Added a comment explaining the `String(v)` coercion is deliberate: OpenAI JSON Schema enums expect string members, while Gemini may supply numeric enum codes. The coercion itself is pre-existing and unchanged.

File: packages/providers/src/openai-shared/schemaConverter.ts [hunk 23]
```diff
@@ -0,0 +1,464 @@
...
+
+  // `required` is handled independently of `properties` so that a nested
+  // object schema declaring `required` without an inline `properties` block
+  // (e.g. one referencing $defs) does not silently drop its required fields.
+  if (Array.isArray(node.required)) {
+    result.required = node.required.map((r) => String(r));
+  } else if (result.type === 'object' && result.properties !== undefined) {
+    result.required = [];
+  }
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/openai-shared/schemaConverter.ts:257]: Valid readability concern. Expanded the existing comment to explicitly state that OpenAI requires `required` to always be present as an array, so object schemas without an explicit `required` default to an empty array. The control flow is unchanged (the empty-array fallback only applies to object-typed nodes with properties), but the rationale is now documented inline.

File: packages/providers/src/openai/schemaConverter.ts [hunk 24]
```diff
@@ -5,50 +5,33 @@
...
-  properties: Record<string, OpenAIPropertySchema>;
-  required: string[];
-  additionalProperties?: boolean;
-  [key: string]: unknown;
-}
+export type {
+  OpenAIFunctionParameters,
+  OpenAIPropertySchema,
+  convertSchemaToOpenAI,
+} from '../openai-shared/schemaConverter.js';
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/openai/schemaConverter.ts:28]: Factually correct that the types are broader than the original classic file, but this broadening is intentional and required by the issue acceptance criteria, not an accidental regression:
(1) `OpenAIPropertySchema` gained an index signature (`[key: string]: unknown`) because the entire point of this refactor is to PRESERVE previously-dropped passthrough keywords (anyOf/oneOf/$ref/format/pattern/const/...). Without the index signature, TypeScript would reject assigning those keys — the signature is load-bearing for the fix.
(2) `additionalProperties?: boolean | OpenAIPropertySchema` reflects that nested `additionalProperties` can legitimately be a schema (the issue explicitly calls out preserving property-level additionalProperties). The original `boolean`-only type was itself incomplete.
The re-export exposes the accurate, richer types. Runtime behavior is unchanged. No action needed.
